weaveworks/weave-gitops
# enhance get session logs to support pod logs, filters, and return logging sources

Fixes #3269 
Fixes #3279
Fixes #3344 

**TODO**
- [x] Get pod logs
- [x] Merged tokens
- [x] Merged logs
- [x] Sorted logs by timestamp
- [x] Extract logging source information
- [x] LogSource and LogLevel Filters
- [x] Return Log sources

Signed-off-by: Chanwit Kaewkasi <chanwit@weave.works>

## Enhance GetSessionLogs

The change of this PR is for a protobuf file in the API core of the project. The file contains the definitions for communication between client and server in the API.

In the `GetSessionLogsRequest` message, two new fields have been added:

- logSourceFilter (string)
- logLevelFilter (string)

In the `GetSessionLogsResponse` message, a new repeated field has been added:

- logSources (repeated string)

The purpose of these changes is to allow filtering of log entries based on log source and log level. The new logSources field in `GetSessionLogsResponse` returns a list of available log sources.

## Implement get pod logs

This PR modifies the function GetSessionLogs in the coreServer struct. The function is used to retrieve logs from either GitOps run logs or pod logs. The changes in this diff do the following:

1. Change the naming of the `virtualClusterName` variable to `clusterName` and handle a special case where the session name is equal to "no-session". If it is, the cluster name is set to the default cluster name.

1. Instead of returning an error on failure, the function returns a `pb.GetSessionLogsResponse` with an error message in the `Error` field.

1. Add the ability to filter logs based on the log source and log level filter. The logs can now be filtered by these values.

1. Add the ability to retrieve logs from the GitOps run logs. If the log source filter is equal to the session log source or empty, the logs will be retrieved from the GitOps run logs.

1. Add the ability to retrieve logs from the pod logs. If the log source filter is not empty, the logs will be retrieved from the pod logs.

1. Sort the logs based on the timestamp before returning the logs in the `pb.GetSessionLogsResponse`.

## Detect log level

This PR is adding a new function called `detectLogLevel` that takes a string as a parameter (representing a log message). The function uses regular expressions to determine if the log message contains keywords indicating the log level, such as "error" or "warning". If it finds a match, it returns a string representation of the log level, either "error", "warn", or "info". If it doesn't find a match, it returns "info" as the default log level.

## Improve getPodLogs

This PR introduces a new struct named `PodLog` that represents the logs data. The struct contains the log data such as `date`, `time`, `log`, `msg`, and `Kubernetes` `metadata` such as pod name, namespace name, pod ID, labels, annotations, host, container name, docker ID, container hash, and container image.

The function `getPodLogs` was modified to read logs data using the new `PodLog` struct. The function uses the Minio client to retrieve log data from an S3 bucket. The logs data is read from the S3 bucket and is then split into individual log entries. The log entries are then transformed into the `PodLog` struct.

If the log entry is empty or has an inner message of an empty string, it is ignored. The log level of each log entry is detected using the `detectLogLevel` function. If a log source filter or log level filter is provided, the logs that do not match the filter are ignored.

The final result of the function is a slice of log entries represented as `pb.LogEntry` objects and a last token representing the last key processed in the S3 bucket.

## Fluent-bit in vcluster

We have encountered a problem with Fluent-Bit in our vcluster environment. The issue is that there is no log directory in vcluster and Fluent-Bit is retrieving information directly from the host's /var/log/containers, which causes a mismatch in container names.

As a result, Fluent-Bit is trying to access information from the vcluster control-plane using the wrong container names (the names at the host level, not the vcluster level). We need to find a solution to this problem to ensure accurate logging in our vcluster environment.

Fortunately, from vcluster v0.13 and later, there is a component called HostPath Mapper to help correct logging information from host so that we can collect them correctly from within vclusters.

For more information.
https://www.vcluster.com/docs/operator/monitoring-logging#logging

## Improve --dashboardImage

This PR also modifies the function `makeValues` in the file `install_dashboard.go`. The main changes in this diff are the addition of a new function `parseImageRepository` and modification of the `makeValues` function to use this new function.

The `parseImageRepository` function takes an input string in the format "repository/image:tag" or "image:tag". The function splits the input string into its component parts (repository, image, and tag) and returns them. If the input string is not in the correct format, the function returns an error.

The `makeValues` function has been updated to call the new `parseImageRepository` function when processing the `dashboardImage` argument. The `dashboardImage` argument is used to specify the Docker image for the GitOps Dashboard. If `dashboardImage` is not an empty string, the function creates a `map[string]interface{}` and sets the image key to a map that contains the repository, image, and tag extracted from the `dashboardImage` argument.

This PR provides a more robust and flexible way of parsing the `dashboardImage` argument and ensures that the input string is in the correct format.

## How to test this PR

- start a WGE instance
- update Weave GitOps dependencies to this branch (`BRANCH=enhance-get-session-logs`)
- go to a local directory of any Git repo you have
- start a GitOps Run session `gitops beta run --skip-dashboard-install ./manifests`
- logs in the GitOps Run UI would appear
